tfx/experimental/distributed_inference/graphdef_experiments/subgraph_partitioning/graph_partition_test.py

```diff
@@ -0,0 +1,107 @@
+# Copyright 2020 Google LLC. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Tests for Graph Partitioning."""
+
+import os
+import tempfile
+import tensorflow as tf
+
+from create_complex_graph import save_examples_as_graphdefs
+from graph_partition import _RemoteOpLayers
+from graph_partition import get_graph_name_to_graph_def, partition_all_graphs
+
+
+class RelationTest(tf.test.TestCase):
+  """Tests for Relation."""
+  def test_layers(self):
+    """Construct an example and validate the layers."""
+    remote_op_relations = {'a1': [], 'a2': [], 'b1': ['a1'],
+                           'b2': ['a1', 'a2'], 'c1': ['b1'],
+                           'c2': ['b1', 'a1', 'b2', 'a2']}
+    desired_outputs = [{'a1', 'a2'}, {'b1', 'b2'}, {'c1', 'c2'}]
+
+    order = _RemoteOpLayers(remote_op_relations)
+    self.assertEqual(desired_outputs, list(order))
+
+
+class PartitionTest(tf.test.TestCase):
+  """Tests for graph partitioning."""
```

Comment:
There is no test that checks for validity of the partitioning algorithm!

For the graphs being used for testing, you should also store a verified "golden" set of partitioned graphs (that you have manually verified to be correct) under a subdir `testdata/` and add a test to verify that the partitioned graphs match the goldens.

Reply:
Added a test that compares the partitioned graphs with a "golden" set. One thing to note: I replaced most sets with lists to make the partitioning algorithm deterministic; the order of traversal is always the same.

Reply:
> One thing to note: I replaced most sets with lists to make the partitioning algorithm deterministic; the order of traversal is always the same.

Can you please revert to `set` where `set` makes sense. Aside from the fact that now you can end up having duplicates in the container (which may be error-prone if not careful), element presence tests would be unnecessarily inefficient. You could instead update the test so that it sorts the nodes or something before it compares.

Reply:
Yes, I reverted back to sets. For the test with golden sets, however, we can't directly compare `GraphDef` protos because sets are not ordered. I moved to comparing the node names for each subgraph layer and added explanations as comments.

Reply:
Thanks